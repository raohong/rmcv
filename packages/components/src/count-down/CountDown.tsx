import classNames from 'classnames';
import { noop, isFunction } from '@rmc-vant/utils';
import React, { useImperativeHandle, useRef, useState } from 'react';
import {
  usePersistFn,
  useUnmountedRef,
  useUpdateEffect,
  useIsomorphicLayoutEffect,
  useUpdateIsomorphicLayoutEffect,
} from '@rmc-vant/hooks';
import { useConfigContext } from '../config-provider';
import CountDownTimer from './Timer';
import type { CountDownRef, CountDownProps, CountDownTimeData } from './interface';
import {
  calCountDownTimeData,
  formatCountDownTimeData,
  santilizeTime,
} from './util';

const CountDown = React.forwardRef<CountDownRef, CountDownProps>(
  (
    {
      className,
      children,
      millisecond,
      time,
      onFinish = noop,
      onChange = noop,
      autoStart = true,
      format = 'HH:mm:ss',
      ...rest
    },
    ref,
  ) => {
    const { getPrefixCls } = useConfigContext();
    const unmountedRef = useUnmountedRef();
    const timer = useRef<CountDownTimer | null>(null);
    const [data, setData] = useState<CountDownTimeData | null>(() => {
      return calCountDownTimeData(autoStart ? santilizeTime(time) : 0);
    });

    const persistedOnChange = usePersistFn(onChange);
    const persistedOnFinish = usePersistFn(onFinish);

    useIsomorphicLayoutEffect(() => {
      const currentTimer = new CountDownTimer(santilizeTime(time), {
        autoStart,
        millisecond,
        onChange: (next) => {
          if (!unmountedRef.current) {
            setData(next);
            persistedOnChange(next);
          }
        },
        onFinish: persistedOnFinish,
      });

      timer.current = currentTimer;

      return () => {
        currentTimer.destory();
        timer.current = null;
      };
    }, []);

    useUpdateIsomorphicLayoutEffect(() => {
      timer.current?.updateConfig(santilizeTime(time), {
        millisecond,
        autoStart,
      });
    }, [time, millisecond, autoStart]);

    useImperativeHandle(ref, () => ({
      start: () => timer.current?.start(),
      pause: () => timer.current?.pause(),
      reset: () => timer.current?.reset(),
    }));

    return (
      <div className={classNames(getPrefixCls('count-down'), className)} {...rest}>
        {
          // eslint-disable-next-line no-nested-ternary
          data !== null
            ? isFunction(children)
              ? children(data)
              : formatCountDownTimeData(data.totalTime, format)
            : null
        }
      </div>
    );
  },
);

export default CountDown;
