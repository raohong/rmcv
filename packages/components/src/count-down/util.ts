import { isNumber } from '@rmc-vant/utils';
import type { CountDownTimeData } from './interface';

const DAY = 3600 * 1000 * 24;

const HOUR = 3600 * 1000;

const MINUTE = 60 * 1000;

const SECOND = 1000;

export const calCountDownTimeData = (time: number): CountDownTimeData => {
  let value = time;

  const days = Math.floor(value / DAY);
  value -= days * DAY;

  const hours = Math.floor(value / HOUR);
  value -= hours * HOUR;

  const minutes = Math.floor(value / MINUTE);
  value -= minutes * MINUTE;

  const seconds = Math.floor(value / SECOND);
  value -= seconds * SECOND;

  const milliseconds = value;

  return {
    days,
    hours,
    minutes,
    seconds,
    milliseconds,
  };
};

const formatMap: Record<string, keyof CountDownTimeData> = {
  D: 'days',
  H: 'hours',
  m: 'minutes',
  s: 'seconds',
};

export const formatCountDownTimeData = (data: CountDownTimeData, format: string) => {
  return format.replace(/[DHms]{1,2}|S{1,3}/g, (matched) => {
    if (matched === undefined) {
      return '';
    }

    if (/^S+$/.test(matched)) {
      return String(data.milliseconds)
        .slice(0, matched.length)
        .padStart(matched.length, '0');
    }

    const key = formatMap[matched[0]];
    const value = String(data[key]);

    return matched.length === 2 ? value.padStart(2, '0') : value;
  });
};

export const santilizeTime = (time?: number) =>
  (isNumber(time) ? Math.max(0, Math.floor(time)) : 0) * 1000;
