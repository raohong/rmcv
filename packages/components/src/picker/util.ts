import { isPlainObject, omit } from '@rmc-vant/utils';
import type {
  PickerBaseOption,
  PickerBaseOptionWithChildren,
  PickerColumns,
  PickerValue,
} from './interface';

export const isNestedOptions = (
  columns: PickerColumns,
): columns is PickerBaseOptionWithChildren[] => {
  if (!columns || !columns[0]) {
    return false;
  }

  return isPlainObject(columns[0]) && 'children' in columns[0];
};

export const getPickerColumnsLength = (columns: PickerColumns) => {
  /**
   * 嵌套结构默认取第一个 item 的 children 深度
   */
  if (isNestedOptions(columns)) {
    let item: PickerBaseOptionWithChildren | undefined = columns[0];
    let length = 0;

    while (item) {
      length++;

      item = item.children?.[0];
    }

    return length;
  }

  return columns.length;
};

export const getPickerColumnData = <V extends PickerValue>(
  value: V[] | undefined,
  columns: PickerColumns<V>,
  length: number,
  fillValue?: boolean,
) => {
  if (!columns || !columns[0]) {
    return {
      columns: [] as PickerBaseOption<V>[][],
      indexList: [] as number[],
    };
  }

  const indexSource = new Array(length).fill(0);

  if (isNestedOptions(columns)) {
    const indexList: number[] = [];
    const result = indexSource.reduce(
      (
        map: {
          source: PickerBaseOptionWithChildren<V>[] | undefined;
          options: PickerBaseOption<V>[][];
        },
        _,
        i,
      ) => {
        if (!map.source) {
          map.options.push([]);

          return map;
        }

        map.options.push(map.source.map((item) => omit(item, ['children'])));

        const targetIndex = Math.max(
          0,
          map.source.findIndex((item) => item.value === value?.[i]),
        );

        if (value && fillValue && value?.[i] === undefined) {
          value[i] = map.source[0].value;
        }

        indexList.push(targetIndex);

        map.source = map.source[targetIndex]?.children;

        return map;
      },
      {
        source: columns,
        options: [],
      },
    );

    return {
      indexList,
      columns: result.options,
    };
  }

  const targetColumns: PickerBaseOption<V>[][] = isPlainObject(columns[0][0])
    ? (columns as PickerBaseOption<V>[][])
    : (columns as V[][]).map((list) =>
        list.map((value) => ({
          value,
          label: value,
        })),
      );

  return {
    columns: targetColumns,
    indexList: indexSource.map((_, i) =>
      Math.max(
        0,
        targetColumns[i].findIndex((item) => item.value === value?.[i]),
      ),
    ),
  };
};

export const fillValueByNestedOptions = <V extends PickerValue>(
  value: V[],
  columns: PickerBaseOptionWithChildren<V>[],
  length: number,
) => {
  const indexSource = new Array(length).fill(0);

  indexSource.reduce(
    (
      map: {
        source: PickerBaseOptionWithChildren<V>[] | undefined;
        options: PickerBaseOption<V>[][];
      },
      _,
      i,
    ) => {
      if (!map.source) {
        map.options.push([]);

        return map;
      }

      map.options.push(map.source.map((item) => omit(item, ['children'])));

      const targetIndex = Math.max(
        0,
        map.source.findIndex((item) => item.value === value?.[i]),
      );

      if (value && value?.[i] === undefined) {
        value[i] = map.source[0].value;
      }

      map.source = map.source[targetIndex]?.children;

      return map;
    },
    {
      source: columns,
      options: [],
    },
  );
};
