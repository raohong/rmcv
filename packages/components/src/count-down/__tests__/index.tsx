import React from 'react';
import { render, screen } from '@testing-library/react';
import { formatCountDownTimeData } from '../util';
import CountDown from '..';

const time = 2000;

test('render correctly', () => {
  const tree = render(<CountDown />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with time and autoStart', () => {
  const format = 'ss';

  render(<CountDown time={time} autoStart format={format} />);

  expect(
    screen.getByText(formatCountDownTimeData(time, format)),
  ).toBeInTheDocument();
});

test('render with render children', () => {
  render(
    <CountDown time={time} autoStart format="s">
      {({ seconds }) => <span data-testid="sec"> {seconds}</span>}
    </CountDown>,
  );

  expect(screen.getByTestId('sec')).toContainHTML('2');
});
