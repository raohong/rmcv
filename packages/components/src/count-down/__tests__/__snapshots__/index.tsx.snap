// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    class="rmcv-count-down"
  >
    00:00:00
  </div>
</DocumentFragment>
`;
